lib/nostrum/shard/session.ex

```diff
@@ -22,50 +27,77 @@ defmodule Nostrum.Shard.Session do
       end
 
     payload = Payload.status_update_payload(idle_since, game, stream, status, afk, type)
-    WebSockex.cast(pid, {:status_update, payload})
+    GenServer.cast(pid, {:status_update, payload})
   end
 
   def request_guild_members(pid, guild_id, limit \\ 0) do
     payload = Payload.request_members_payload(guild_id, limit)
-    WebSockex.cast(pid, {:request_guild_members, payload})
+    GenServer.cast(pid, {:request_guild_members, payload})
   end
 
   def start_link([gateway, shard_num]) do
+    GenServer.start_link(__MODULE__, [gateway, shard_num])
+  end
+
+  def init([_gateway, _shard_num] = args) do
+    {:ok, nil, {:continue, args}}
+  end
+
+  def handle_continue([gateway, shard_num], nil) do
+    Connector.block_until_connect()
+    Logger.metadata(shard: shard_num)
+
+    worker = connect(gateway)
+
+    zlib_context = :zlib.open()
+    :zlib.inflateInit(zlib_context)
+
     state = %WSState{
+      conn_pid: self(),
+      conn: worker,
       shard_num: shard_num,
       gateway: gateway <> @gateway_qs,
       last_heartbeat_ack: DateTime.utc_now(),
-      heartbeat_ack: true
+      heartbeat_ack: true,
+      zlib_ctx: zlib_context
     }
 
-    Connector.block_until_connect()
+    Logger.debug(fn -> "Websocket connection up on worker #{inspect(worker)}." end)
 
-    # TODO: Add support for `spawn_opt` start arguments to WebSockex, this does nothing until then.
-    WebSockex.start_link(
-      gateway <> @gateway_qs,
-      __MODULE__,
-      state,
-      spawn_opt: [Util.fullsweep_after()]
-    )
+    {:noreply, state}
   end
 
-  def handle_connect(conn, state) do
-    Logger.metadata(shard: state.shard_num)
+  @spec connect(String.t()) :: pid()
+  defp connect(gateway) do
```

Comment:
I'm not sure this needs to be in its own function, but if you really want it to I'm fine with that. We're only calling it in one place.

Reply:
Leftover from old reconnection code. Dropped like I drop my ubers